### DISCUSSION TITLE
[NativeAOT] How do I link a static library from a nuget package?

### DISCUSSION THREAD

**Discussion by slxdy** [dotnet/runtime#108981] [Q&A] [answered]
In my NativeAOT project, I would like the link the static libraries from `Microsoft.NETCore.DotNetAppHost`.
Is this possible? If yes, how?

**Comment by neon-sunset** [comment]
It really comes down to specifying DirectPInvoke and NativeLibrary msbuild properties. For specific integrations consumed by .NET itself like ICU - there are separate properties (see https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/compiling.md).

Which library do you want to link to?

**Reply by slxdy** [reply]
I want to link `libnethost.lib` from the `Microsoft.NETCore.DotNetAppHost` nuget package. (Also for some reason, for win-x64 the package contains both `nethost.lib` and `libnethost.lib`, though the latter is much bigger)

**Reply by AustinWise** [reply]
With regards to choosing between `nethost.lib` and `libnethost.lib`, it depends on whether you want to statically or dynamically link the nethost library. If you link against the smaller `nethost.lib` you binary will dynamically link against `nethost.dll`. If you use the larger `libnethost.lib`, the contents of the nethost library will be statically linked into your binary.

**Reply by slxdy** [reply]
Alright thanks, makes sense. Now I just gotta figure out how to link from the nuget package

**Reply by slxdy** [reply]
To clarify, I know how to link static libs, but I'd like to do so from the nuget package. I don't want to include the lib in my repo

**Comment by AustinWise** [comment, marked as answer]
I wrote a sample that statically links the `nethost` API into NativeAOT app. This sample currently only works when published as NativeAOT. If I get more time, I may extend it support dynamic linking so that it can be run as a normal .NET app on CoreCLR.

https://github.com/AustinWise/DotNetHostingAndNativeAot

Drawing on the [sample app](https://github.com/dotnet/samples/blob/main/core/hosting/src/NativeHost/NativeHost.csproj), instead of using the `Microsoft.NETCore.DotNetAppHost` package to get the `nethost` library, this app finds the `nethost` library from the targeting pack. Typically the .NET SDK includes this.

I tested this solution on Windows 11, Ubuntu 24.04, and macOS 15. On Linux, you might have to do something like `apt-get install libstdc++-dev`.

I'll reproduce the current state of my solution below:

## Project file

```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net8.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
        <PublishAot>true</PublishAot>
        <InvariantGlobalization>true</InvariantGlobalization>
        <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    </PropertyGroup>

    <Target Name="PrepareForNativeBuild" AfterTargets="Build">
        <PropertyGroup>
            <NetHostDir>$(NetCoreTargetingPackRoot)/Microsoft.NETCore.App.Host.$(NETCoreSdkRuntimeIdentifier)/$(BundledNETCoreAppPackageVersion)/runtimes/$(NETCoreSdkRuntimeIdentifier)/native</NetHostDir>
            <NetHostName Condition="$([MSBuild]::IsOsPlatform('Windows'))">libnethost.lib</NetHostName>
            <NetHostName Condition="!$([MSBuild]::IsOsPlatform('Windows'))">libnethost.a</NetHostName>
        </PropertyGroup>

        <ItemGroup>
            <NativeLibrary Include="$(NetHostDir)/$(NetHostName)" />
        </ItemGroup>

        <ItemGroup Condition="!$([MSBuild]::IsOsPlatform('Windows'))">
            <!-- This is needed by libnethost.a. -->
            <LinkerArg Include="-lstdc++" />
        </ItemGroup>
    </Target>

</Project>
```

## Code file

This prints the host FXR path:

```csharp
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;

[assembly: DisableRuntimeMarshalling]

partial class Program
{
    static void Main()
    {
        try
        {
            string hostFxrPath = GetHostFxrPath();
            Console.WriteLine($"Got host FXR Path: {hostFxrPath}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Failed to get host FXR path: {ex.Message}");
        }
    }

    private static string GetHostFxrPath()
    {
        if (OperatingSystem.IsWindows())
        {
            char[] buffer = new char[1024];
            nint bufferSize = buffer.Length;
            int rc = get_hostfxr_path_windows(buffer, ref bufferSize, IntPtr.Zero);
            if (rc == 0)
            {
                return new string(buffer.AsSpan().Slice(0, (int)bufferSize - 1));
            }
            else
            {
                throw new Exception($"get_hostfxr_path failed: {rc}");
            }
        }
        else
        {
            byte[] buffer = new byte[1024];
            nint bufferSize = buffer.Length;
            int rc = get_hostfxr_path_non_windows(buffer, ref bufferSize, IntPtr.Zero);
            if (rc == 0)
            {
                return Encoding.UTF8.GetString(buffer.AsSpan().Slice(0, (int)bufferSize - 1));
            }
            else
            {
                throw new Exception($"get_hostfxr_path failed: {rc}");
            }
        }
    }

    [LibraryImport("*", EntryPoint = "get_hostfxr_path")]
    private static partial int get_hostfxr_path_windows(char[] buffer, ref nint buffer_size, IntPtr parameters);

    [LibraryImport("*", EntryPoint = "get_hostfxr_path")]
    private static partial int get_hostfxr_path_non_windows(byte[] buffer, ref nint buffer_size, IntPtr parameters);
}
```

**Reply by slxdy** [reply]
That is... really smart. Didn't think of getting it from the dotnet sdk itself lol. Thanks a lot
